File: services/py/src/infrastructure/http/network/forwarder.py
from src.domain.entities.request import RequestDTO
from src.infrastructure.http.network.stream import Stream
from src.infrastructure.http.request.formater import RequestFormatter


class HttpForwarder:

    @staticmethod
    async def forward(
        request: RequestDTO,
        client: Stream,
        upstream: Stream,
        client_ip: str
    ) -> None:
        headers = RequestFormatter.format_headers(request, client_ip)
        await HttpForwarder._forward_headers(request, upstream, headers)
        await HttpForwarder._forward_body(headers, client, upstream)

    @staticmethod
    async def _forward_headers(request: RequestDTO, upstream: Stream, headers) -> None:
        new_line = f"{request.build_line()}\r\n"
        await upstream.write(data=new_line.encode())
        for k, v in headers.items():
            await upstream.write(data=f"{k}: {v}\r\n".encode())

        await upstream.write(data=b"\r\n", is_drain=True)

    @staticmethod
    async def _forward_body(headers: dict[str, str], client: Stream, upstream: Stream) -> None:
        if "content-length" in headers:
            remaining = int(headers["content-length"])
            while remaining > 0:
                chunk = await client.read(min(8192, remaining))
                if not chunk:
                    break
                await upstream.write(chunk, is_drain=True)
                remaining -= len(chunk)

        elif headers.get("transfer-encoding") == "chunked":
            while True:
                line = await client.readline()
                if not line:
                    break

                await upstream.write(line, is_drain=True)
                size = int(line.strip(), 16)
                if size == 0:
                    break

                chunk = await client.read_exactly(size + 2)
                await upstream.write(chunk, is_drain=True)

File: services/go/src/internal/infrastructure/network/forwarder.go
package network

import (
	"fmt"
	"strconv"
)

func Forward(req *Request, client *Stream, upstream *Stream, clientIP string) error {
	headers := FormatHeaders(req, clientIP)
	if err := upstream.Write([]byte(req.BuildLine()+"\r\n"), false); err != nil {
		return err
	}
	for k, v := range headers {
		if err := upstream.Write([]byte(fmt.Sprintf("%s: %s\r\n", k, v)), false); err != nil {
			return err
		}
	}
	if err := upstream.Write([]byte("\r\n"), true); err != nil {
		return err
	}

	if cl, ok := headers["content-length"]; ok {
		n, _ := strconv.Atoi(cl)
		remaining := n
		bufSize := 8192
		for remaining > 0 {
			toRead := bufSize
			if remaining < bufSize {
				toRead = remaining
			}
			chunk, err := client.Read(toRead)
			if err != nil {
				return err
			}
			if len(chunk) == 0 {
				break
			}
			if err := upstream.Write(chunk, true); err != nil {
				return err
			}
			remaining -= len(chunk)
		}
	} else if te, ok := headers["transfer-encoding"]; ok && te == "chunked" {
		for {
			line, err := client.ReadLine()
			if err != nil {
				return err
			}
			if err := upstream.Write(line, true); err != nil {
				return err
			}
			// parse chunk size
			sizeLine := string(line)
			size, _ := strconv.ParseInt(sizeLine, 16, 64)
			if size == 0 {
				rest, _ := client.ReadLine()
				upstream.Write(rest, true)
				break
			}
			chunk, err := client.ReadExactly(int(size) + 2) // +CRLF
			if err != nil {
				return err
			}
			if err := upstream.Write(chunk, true); err != nil {
				return err
			}
		}
	}
	return nil
}
